### PR TITLE
Fix DeepSeek routing logic

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

* Add device-limit logic into DeepSeek routing
* Add a `GateLogit` layer to return pre_bias_output

**I will have another PR to move MoE related blocks out of linears.py into moe.py**

# Tests

* github runners pass
* test smaller config of v2 without sharding issue with `pre_bias_logits_pspec = None` [link](https://paste.googleplex.com/6226575585181696)
* test on v3 logits comparison, and see improvement [link](https://b.corp.google.com/issues/394669494#comment39)
* Reload the v2-lite ckpt, and no issues for decoding 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/layers/linears.py [hunk 1]
```diff
@@ -988,7 +1028,7 @@ def retrieve_quantized_weight(
   def __call__(self, inputs):
     cfg = self.config
     inputs = inputs.astype(cfg.dtype)
-    gate_logits = DenseGeneral(
+    gate_logits, pre_bias_logits = DenseGeneral(
```
Comment (gagika) [inline MaxText/layers/linears.py:1031]: instead of modifying DenseGeneral to fit DeepSeek's need, let's create some new layer (perhaps inheriting from DenseGeneral) with DeepSeek specific change, e.g. pre_bias_logits logic and score_func=self.config.routed_score_func, wdyt?

Comment (RissyRan) [reply MaxText/layers/linears.py:1031]: That's a good point. Will update the change.

Comment (RissyRan) [reply MaxText/layers/linears.py]: Hi Gagik, updated to a new module named as `DenseWithNormBias` as `use_bias` is also used in mlp block and gpt3 block, so not DeepSeek specific. Feel free to let me know your thoughts! Thank you!

File: MaxText/layers/linears.py [hunk 2]
```diff
@@ -93,8 +93,6 @@ class DenseGeneral(nn.Module):
     weight_dtype: the dtype of the weights (default: float32).
     dtype: the dtype of the computation (default: float32).
     kernel_init: initializer function for the weight matrix.
-    use_bias: whether to add bias in linear transformation.
-    score_func: scoring function for output normalization.
```
Comment (gagika) [inline MaxText/layers/linears.py:97]: the common convention for Dense layer is to have linear layer and bias, 
e.g. flax.linen.Dense, keras.layers.Dense, torch.nn.Linear
Let's follow the convention for DenseGeneral too, unless we have a strong reason not too, and gpt3 can use it as is.

We can remove score_func from DenseGeneral, as it's very DeepSeek specific.

File: MaxText/layers/linears.py [hunk 3]
```diff
@@ -156,11 +152,36 @@ def compute_dot_general(inputs, kernel, axis, contract_ind):
 
     contract_ind = tuple(range(0, len(axis)))
     output = compute_dot_general(inputs, kernel, axis, contract_ind)
+    return output
+
+
+class DenseWithNormBias(DenseGeneral):
```
Comment (gagika) [inline MaxText/layers/linears.py:158]: If we are going to have a separate layer Let's make this layer very specific for DeepSeek Gating (e.g. reflecting in the name).

We can also not have a separate layer, and for DeepSeek use `DenseGeneral` for DeepSeek, then  add a bias (while keeping previous weights) and apply scaling.

Both options are good in my opinion.

**Review by gagika — COMMENTED** [review]
a few comments about the design of the layer.

File: MaxText/layers/linears.py (hunk 3 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:158]: Yeah, sounds good. The tricky thing is this is just a dense matmul, not like gating, which also need to combines routing logic. I think we have 2 good options, 1) have 'DenseDeepseek', which will have some duplicate logics with 'DenseGeneral', as score_func change is in the middle (before bias). 2) (this could be a later version) refactor routing logics into a new 'MoeGate', this will need ckpt structure changed (probably just an extra layer), since MoeGate will call DenseGeneral or DenseDeepSeek and then select topk. I will do a refactor option 1 first, and follow up with 2nd option.

Comment (RissyRan) [reply MaxText/layers/linears.py:158]: Let's quickly chat in person later? and then I will do the refactor once we are aligned. Thank you!

File: MaxText/configs/base.yml [hunk 4]
```diff
@@ -147,6 +147,8 @@ shared_experts: 1
 routed_scaling_factor: 1.0 # scaling factor for routing scores
 routed_score_func: "" # scoring function for routing
 routed_bias: False # a flag if a bias term is added for routing
+n_groups: 1 # number of groups for routing
```
Comment (gobbleturk) [inline MaxText/configs/base.yml:150]: can we give this a more specific name like n_routing_groups

Comment (RissyRan) [reply MaxText/configs/base.yml:150]: Sounds good! Will update both to `n_routing_groups` and `topk_routing_group`

File: MaxText/layers/linears.py (hunk 2 above)
Comment (RissyRan) [reply MaxText/layers/linears.py]: Discussed offline, and updated accordingly.

File: MaxText/layers/linears.py [hunk 5]
```diff
@@ -106,7 +105,99 @@ class DenseGeneral(nn.Module):
...
+          bias_shape,
+          self.weight_dtype,
+      )
+      bias = jnp.asarray(bias, self.dtype)
+      output += bias
+    return output
+
+
+class GateLogit(nn.Module):
+  """A linear transformation with flexible axes.
```
Comment (gagika) [inline MaxText/layers/linears.py:175]: could you update this description?

This layer is used gate logit calculation, allowing for the specific pre-bias logic needed for DeepSeek routing.

File: MaxText/layers/linears.py [hunk 6]
```diff
@@ -106,7 +105,99 @@ class DenseGeneral(nn.Module):
...
+
+  Attributes:
+    features: tuple with numbers of output features.
+    decoder_block: which DecoderBlock to use.
+    axis: tuple with axes to apply the transformation on.
+    weight_dtype: the dtype of the weights (default: float32).
+    dtype: the dtype of the computation (default: float32).
+    kernel_init: initializer function for the weight matrix.
+    kernel_axes: tuple with axes to apply kernel function.
+    use_bias: whether to add bias in linear transformation.
```
Comment (gagika) [inline MaxText/layers/linears.py:185]: Could you add something like this: 
whether to add bias in gate logit scores, it's not part of loss function, but used for expert load balancing (similar to DeepSeek v3)

File: MaxText/layers/linears.py [hunk 7]
```diff
@@ -156,9 +247,13 @@ def compute_dot_general(inputs, kernel, axis, contract_ind):
 
     contract_ind = tuple(range(0, len(axis)))
     output = compute_dot_general(inputs, kernel, axis, contract_ind)
+    pre_bias_logits = None
 
     if self.score_func:
       output = _convert_to_activation_function(self.score_func)(output)
+      if self.decoder_block == "deepseek" and self.use_bias:
```
Comment (gagika) [inline MaxText/layers/linears.py:254]: Could you check deepseek v3? e.g. config.model_name.startswith("deepseek3")

File: MaxText/layers/linears.py [hunk 8]
```diff
@@ -382,17 +490,47 @@ def deepseek_scale_weights(self, weights):
     weights *= self.config.routed_scaling_factor
     return weights
 
-  def permute(self, inputs, gate_logits):
+  def deepseek_routing(self, gate_logits, pre_bias_logits):
+    """DeepSeek routing logit."""
```
Comment (gagika) [inline MaxText/layers/linears.py:494]: Some comments here?

e.g. 
This involves a two-stage selection: first selecting top groups of experts,
then selecting the top experts within those chosen groups. Selection uses
post-bias/post-score_func logits, but the returned weights are based on
pre-bias/pre-score_func logits.

File: MaxText/layers/linears.py [hunk 9]
```diff
@@ -382,17 +490,47 @@ def deepseek_scale_weights(self, weights):
     weights *= self.config.routed_scaling_factor
     return weights
 
-  def permute(self, inputs, gate_logits):
+  def deepseek_routing(self, gate_logits, pre_bias_logits):
```
Comment (gagika) [inline MaxText/layers/linears.py:493]: Can this be disabled, if users want to run with different routing?

**Review by gagika — APPROVED** [review]
Thanks, only minor comments

File: MaxText/layers/linears.py (hunk 5 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:175]: SG. Updated to `A layer used to compute gate logits, allowing to return the pre bias values for DeepSeek routing.`

File: MaxText/layers/linears.py (hunk 6 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:185]: Updated to `use_bias: whether to add learnable bias in gate logit scores. When enabled, this bias aids expert load balancing (like in DeepSeek V3) and is not part of the loss calculation.

File: MaxText/layers/linears.py (hunk 7 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:254]: sounds good, will update

File: MaxText/layers/linears.py (hunk 9 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:493]: Add a TODO `TODO(ranran): Add an option to turn off device-limited routing

File: MaxText/layers/linears.py [hunk 10]
```diff
@@ -157,8 +155,96 @@ def compute_dot_general(inputs, kernel, axis, contract_ind):
...
+  kernel_axes: Tuple[Optional[str], ...] = ()
+  use_bias: bool = False
+  score_func: str = ""
+  quant: Optional[Quant] = None
+  matmul_precision: str = "default"
+
+  @nn.compact
+  def __call__(self, inputs: Array) -> Array:
+
+    def compute_dot_general(inputs, kernel, axis, contract_ind):
```
Comment (gobbleturk) [inline MaxText/layers/linears.py:206]: is there a way we can share code with DenseGeneral? I assume this was copied from DenseGeneral?

Comment (RissyRan) [reply MaxText/layers/linears.py:206]: Yes, I could refactor this out as a global helper function. Will update that

File: MaxText/configs/base.yml [hunk 11]
```diff
@@ -147,6 +147,8 @@ shared_experts: 1
 routed_scaling_factor: 1.0 # scaling factor for routing scores
 routed_score_func: "" # scoring function for routing
 routed_bias: False # a flag if a bias term is added for routing
+n_routing_groups: 1 # number of groups for routing
+topk_routing_group: 1 # number of top groups to route inputs
```
Comment (gobbleturk) [inline MaxText/configs/base.yml:151]: Is there a recommendation for how this interacts with EP? My guess is that this only can increase performance when topk_routing_groups < EP

Perhaps some comment here about performance tradeoff? 
"The activations only have to be sent to at most topk_routing_group different devices for EP, which can have performance benefits"

File: MaxText/layers/linears.py [hunk 12]
```diff
@@ -173,7 +259,7 @@ def compute_dot_general(inputs, kernel, axis, contract_ind):
       )
       bias = jnp.asarray(bias, self.dtype)
       output += bias
-    return output
+    return output, pre_bias_logits
 
 
 class MlpBlock(nn.Module):
```
Comment (gobbleturk) [inline MaxText/layers/linears.py:265]: I wonder if we can have a separate dense MlpBlock versus sparse... Don't have to fix in this PR....

File: MaxText/layers/linears.py [hunk 13]
```diff
@@ -382,17 +481,56 @@ def deepseek_scale_weights(self, weights):
     weights *= self.config.routed_scaling_factor
     return weights
 
-  def permute(self, inputs, gate_logits):
-    """Permute tokens to group by expert to fit gmm call."""
+  def deepseek_routing(self, gate_logits, pre_bias_logits):
+    """DeepSeek routing logit. It involves two-stage selection process:
 
+    1) Group Scoring: Experts are partitioned into n_routing_groups.
```
Comment (gobbleturk) [inline MaxText/layers/linears.py:487]: Great comment!

**Review by gobbleturk — CHANGES_REQUESTED** [review]
Mostly LGTM! Can we add a unit test for the new logic for the new device limited `deepseek_routing`?

File: MaxText/configs/models/deepseek3-671b.yml [hunk 14]
```diff
@@ -32,6 +32,8 @@ shared_experts: 1
 routed_scaling_factor: 2.5
 routed_score_func: "sigmoid"
 routed_bias: True
+n_routing_groups: 8
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/configs/models/deepseek3-671b.yml:35]: Yes, that's right! I will add a comment `adjustable to controlling communication costs in routing` for now, and will update it according to some benchmarks later this week (i.e. values better for TPUs).

File: MaxText/layers/linears.py (hunk 12 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:265]: What will be the delta for sparseMlpBlock?

File: MaxText/layers/linears.py [hunk 15]
```diff
@@ -382,17 +481,56 @@ def deepseek_scale_weights(self, weights):
     weights *= self.config.routed_scaling_factor
     return weights
 
-  def permute(self, inputs, gate_logits):
-    """Permute tokens to group by expert to fit gmm call."""
+  def deepseek_routing(self, gate_logits, pre_bias_logits):
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/layers/linears.py:484]: Sure, will add.

File: MaxText/configs/base.yml (hunk 11 above)
Comment (RissyRan) [reply MaxText/configs/base.yml:151]: Yes, topk_routing_groups < EP, and it should improve the all_to_all communication. However, how GPU structure is different to TPU, so like you also mentioned, this config could be configuable, or we could even have an option to turn it off. I am planning to have some tests later this week. 

Will have the comment you mentioned above. Thanks!

File: MaxText/configs/models/deepseek3-671b.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (gobbleturk) [reply MaxText/configs/models/deepseek3-671b.yml]: Can we instead remove these from /configs/model/deepseek? Otherwise allall users of deepseek are forced to use exactly these values - we have logic that enforces config set in /configs/models are not overriden by CLI, but users may want to override these particular `n_routing_groups` choices

See our logic for not overwriting the /config/model settings:
https://github.com/AI-Hypercomputer/maxtext/blob/948fb88eceb0a35167f43fa190436649e7b2522e/MaxText/pyconfig.py#L243-L249

File: MaxText/layers/linears.py (hunk 12 above)
Comment (gobbleturk) [reply MaxText/layers/linears.py:257]: like 400 lines of code or so lol

**Review by gobbleturk — APPROVED** [review]
